### PR TITLE
Add reducer 2 to implement 1

### PR DESCRIPTION
**Pull request by schettn**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently there is no way to interact with pure SQL over models. #8
- Currently there is no OSM reconstructor. #9
- Currently there are no abstract statement objects. #10

**What is the new behavior (if this is a feature change)?**
- Now there is a way to interact with pure SQL over models. #8
- Now there is an OSM reconstructor. #9
- Now there are abstract statement objects. #10

**Other information**:

- Ref: https://github.com/agershun/alasql
- Ref: https://momentjs.com/docs/

### CODE REVIEW

File: src/reducer/database/helper/general.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Reduce object fields by using a key schema
+function squeezer(obj, keys) {
+  //let t = Object.keys(obj).concat(keys)
```
Comment (pinterid) [inline src/reducer/database/helper/general.js]: Remove!

File: src/reducer/database/helper/statistic.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import moment from "moment";
+
+function calculateStreaks(values) {
+  if (!values) {
```
Comment (pinterid) [inline src/reducer/database/helper/statistic.js]: What is the use case of this, can't you just return an empty list if the values are empty?

File: src/reducer/database/models.js [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+  getStreaks() {
+    if (this.year || this.year === 0) {
+
+      let { from, to } = this.getDates();
+      let response = Calendar.getDaysBetweenDate(this, {
+        from,
+        to
+      });
+
+      // response.push({date: "2019-12-01", total:5})
```
Comment (pinterid) [inline src/reducer/database/models.js]: Remove!

File: src/reducer/database/osm/reconstructor.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      }
+    };
+    this.all = (...fields) => {
+      try {
+        let response;
+        response = SOAssambler.database.exec(Base.statements.all, fields);
+        response = response.map((entry) => {
+          return new Base(entry);
+        });
+
```
Comment (pinterid) [inline src/reducer/database/osm/reconstructor.js]: Remove empty line!

File: src/reducer/database/osm/reconstructor.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+          success: false,
+          response: null,
+          message: err.message
+        };
+      }
+    };
+    this.custom = (query) => {
+      try {
+        let response = SOAssambler.database.exec(query);
+
```
Comment (pinterid) [inline src/reducer/database/osm/reconstructor.js]: Remove empty line !

File: src/reducer/database/models.js [hunk 6]
```diff
@@ -0,0 +1,461 @@
+import * as osm from "./osm";
+import moment, { lang } from "moment";
+// import * as helper from "./helper";
```
Comment (pinterid) [inline src/reducer/database/models.js]: Remove!

File: src/reducer/database/osm/models.js [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+class PlatformSO extends BaseSO {
+  static statements = platform;
+
+  constructor() {
+    super();
+  }
+  createRepository(fields) {}
+  createOrganization(fields) {}
+  createStatistic(fields) {}
+
```
Comment (pinterid) [inline src/reducer/database/osm/models.js]: Missing createCalendarEntry!

File: src/reducer/database/osm/models.js [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+  getOrganizations(fields) {}
+  getRepositories(fields) {}
+  getStatistics(fields) {}
+
```
Comment (pinterid) [inline src/reducer/database/osm/models.js]: Missing getCalendar & getActivity!

File: src/reducer/database/osm/models.js [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+import * as statistic from "./statements/statistic";
+import * as streak from "./statements/streak";
+import * as calendar from "./statements/calendar";
+import * as contribution from "./statements/contribution";
+
+// Statement Objects => SO
+import * as helper from "../helper";
+
+class BaseSO {
+  //static objects = SOAssambler;
```
Comment (pinterid) [inline src/reducer/database/osm/models.js]: Remove!

File: src/reducer/database/osm/statements/calendar.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  SELECT 
+      total,
+      date,
+      platform_id
+  FROM 
+      calendar
+  WHERE 
+      total = (
+          SELECT 
+              MAX(total )
```
Comment (pinterid) [inline src/reducer/database/osm/statements/calendar.js]: Remove space afrer total.

File: src/reducer/database/osm/statements/member.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    id=?
+`;
+
+const all = `
+  SELECT
+    *
+  FROM
+    member
+`;
+
```
Comment (pinterid) [inline src/reducer/database/osm/statements/member.js]: Remove empty lines!

File: src/reducer/database/osm/models.js [hunk 12]
```diff
@@ -0,0 +1,703 @@
...
+    //> Usage
+    // Get merged calendar
+
+    //> Implementation example
+    //Calendar.getCalendar(dates);
+  }
+}
+
+/**
+ * Comment
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: Please replace all placeholder comments.

File: src/reducer/index.js [hunk 13]
```diff
@@ -1,18 +1,28 @@
-import { Database } from "./database";
-import { Reducer as SampleReducer } from "./sample.js";
+import * as models from "./database/models";
+import * as helper from "./helper";
```
Comment (kleberbaum) [inline src/reducer/index.js]: Pleas add a "Class" headline.

File: src/reducer/database/models.js [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+
+    this.id = args["id"];
+    this.color = args["color"];
+    this.name = args["name"];
+    this.size = args["size"];
+    this.share = args["share"];
+    this.repositoryId = args["repositoryId"];
+  }
+
+  //> Static methods
```
Comment (kleberbaum) [inline src/reducer/database/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 15]
```diff
@@ -0,0 +1,708 @@
...
+      PlatformSO.statements.lowestCreatedAtYear
+    )[0];
+  }
+
+  //> Abstract Methods
+  createRepository(fields) {
+    //> Usage
+    // Creates a repository within a platform.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 16]
```diff
@@ -0,0 +1,708 @@
...
+    //);
+
+    //return repository;
+  }
+
+  createOrganization(fields) {
+    //> Usage
+    // Creates a organization within a platform.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 17]
```diff
@@ -0,0 +1,708 @@
...
+    //);
+
+    //return organization;
+  }
+
+  createStatistic(fields) {
+    //> Usage
+    // Creates a statistic within a platform.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 18]
```diff
@@ -0,0 +1,708 @@
...
+    //let statistic = Statistic.objects.create(fields);
+
+    //return statistic;
+  }
+
+  createCalendarEntry(fields) {
+    //> Usage
+    // Creates a calendar entry within a platform.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 19]
```diff
@@ -0,0 +1,708 @@
...
+    //let calendar = Calendar.objects.create(fields);
+
+    //return calendar;
+  }
+
+  getRepositories() {
+    //> Usage
+    // Get all repositories within a platform.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 20]
```diff
@@ -0,0 +1,708 @@
...
+    //);
+
+    //return repositories;
+  }
+
+  getOrganizations() {
+    //> Usage
+    // Get all organizations within a platform.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 21]
```diff
@@ -0,0 +1,708 @@
...
+    //);
+
+    //return organizations;
+  }
+
+  getCalendar(dates) {
+    //> Usage
+    // Get merged calendar
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 22]
```diff
@@ -0,0 +1,708 @@
...
+    let response = cls.objects.custom(language.byRepository(self.id));
+    return response;
+  }
+
+  //> Abstract Methods
+  createMember(fields) {
+    //> Usage
+    // Create a member within a repository.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 23]
```diff
@@ -0,0 +1,708 @@
...
+    //  memberId: member.id
+    //});
+
+    //return member;
+  }
+  createLanguage(fields) {
+    //> Usage
+    // Create a languages within a repository.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 24]
```diff
@@ -0,0 +1,708 @@
...
+    //let language = Language.objects.create(fields);
+
+    //return language;
+  }
+
+  getMembers() {
+    //> Usage
+    // Get all members of repository.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 25]
```diff
@@ -0,0 +1,708 @@
...
+
+    return response;
+  }
+
+  //> Abstract Methods
+  createMember(fields) {
+    //> Usage
+    // Create a member within a repository.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 26]
```diff
@@ -0,0 +1,708 @@
...
+    //});
+
+    //return member;
+  }
+
+  getMembers(fields) {
+    //> Usage
+    // Get all members of a repository.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 27]
```diff
@@ -0,0 +1,708 @@
...
+
+    return response;
+  }
+
+  //> Abstract Methods
+  createStreak(fields) {
+    //> Usage
+    // Create a streak within a statistic.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 28]
```diff
@@ -0,0 +1,708 @@
...
+    //let streak = Streak.objects.create(fields);
+
+    //return streak;
+  }
+
+  getStreaks() {
+    //> Usage
+    // Get all streaks of a statistic.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 29]
```diff
@@ -0,0 +1,708 @@
...
+    //  return response;
+    //}
+    //return [];
+  }
+
+  getBusiestDay() {
+    //> Usage
+    // Get the busiest day between the result of getDates().
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 30]
```diff
@@ -0,0 +1,708 @@
...
+
+    //  return response;
+    //}
+  }
+
+  getDates(){
+    //> Usage
+    // Get the correct from and to date based on the year.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 31]
```diff
@@ -0,0 +1,708 @@
...
+    //  from,
+    //  to
+    //};
+  }
+  
+  getStreakDetails(streaks){
+    //> Usage
+    // Get the current and longest streak of streaks.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/database/osm/models.js [hunk 32]
```diff
@@ -0,0 +1,708 @@
...
+      };
+    }
+  }
+
+  //> Abstract Methods
+  createContribution(fields) {
+    //> Usage
+    // Create a contribution within a calendar entry.
+
+    //> Implementation example
```
Comment (kleberbaum) [inline src/reducer/database/osm/models.js]: All headlines should be in **title case**.

File: src/reducer/helper/calendar.js [hunk 33]
```diff
@@ -0,0 +1,60 @@
+//> Moment
+// A lightweight JavaScript date library for parsing, validating, manipulating, and formatting dates.
+import moment from "moment";
+
+//> Database models
```
Comment (kleberbaum) [inline src/reducer/helper/calendar.js]: All headlines should be in **title case**.

File: src/reducer/helper/language.js [hunk 34]
```diff
@@ -0,0 +1,23 @@
+//> Database models
```
Comment (kleberbaum) [inline src/reducer/helper/language.js]: All headlines should be in **title case**.

File: src/reducer/helper/profile.js [hunk 35]
```diff
@@ -0,0 +1,54 @@
+//> Database models
```
Comment (kleberbaum) [inline src/reducer/helper/profile.js]: All headlines should be in **title case**.

File: src/reducer/helper/statistic.js [hunk 36]
```diff
@@ -0,0 +1,49 @@
+//> Database models
```
Comment (kleberbaum) [inline src/reducer/helper/statistic.js]: All headlines should be in **title case**.

File: src/reducer/index.js [hunk 37]
```diff
@@ -1,18 +1,32 @@
-import { Database } from "./database";
-import { Reducer as SampleReducer } from "./sample.js";
+//> Database models
```
Comment (kleberbaum) [inline src/reducer/index.js]: All headlines should be in **title case**.

**Pull request merged**